container/apptabs.go

```diff
@@ -3,678 +3,312 @@ package container
+func (t *AppTabs) SetTabLocation(l TabLocation) {
+	// Mobile has limited screen space, so don't put app tab bar on long edges
+	if d := fyne.CurrentDevice(); d.IsMobile() {
+		if o := d.Orientation(); fyne.IsVertical(o) {
+			if l == TabLocationLeading || l == TabLocationTrailing {
+				l = TabLocationBottom
+			}
+		} else {
+			if l == TabLocationTop || l == TabLocationBottom {
+				l = TabLocationLeading
+			}
 		}
 	}
-
-	return mismatch
+	t.baseTabs.SetTabLocation(l)
 }
 
-type appTabsRenderer struct {
-	animation       *fyne.Animation
-	container       *AppTabs
-	tabLoc          TabLocation
-	line, underline *canvas.Rectangle
-	objects         []fyne.CanvasObject // holds only the CanvasObject of the tabs' content
-	tabBar          *fyne.Container
-}
+// Declare conformity with WidgetRenderer interface.
+var _ fyne.WidgetRenderer = (*appTabsRenderer)(nil)
 
-func (r *appTabsRenderer) Destroy() {
+type appTabsRenderer struct {
+	baseTabsRenderer
+	appTabs *AppTabs
 }
 
 func (r *appTabsRenderer) Layout(size fyne.Size) {
-	tabBarMinSize := r.tabBar.MinSize()
-	var tabBarPos fyne.Position
-	var tabBarSize fyne.Size
-	var linePos fyne.Position
-	var lineSize fyne.Size
-	var childPos fyne.Position
-	var childSize fyne.Size
-	switch r.adaptedLocation() {
-	case TabLocationTop:
-		buttonHeight := tabBarMinSize.Height
-		tabBarPos = fyne.NewPos(0, 0)
-		tabBarSize = fyne.NewSize(size.Width, buttonHeight)
-		linePos = fyne.NewPos(0, buttonHeight)
-		lineSize = fyne.NewSize(size.Width, theme.Padding())
-		barHeight := buttonHeight + theme.Padding()
-		childPos = fyne.NewPos(0, barHeight)
-		childSize = fyne.NewSize(size.Width, size.Height-barHeight)
-	case TabLocationLeading:
-		buttonWidth := tabBarMinSize.Width
-		tabBarPos = fyne.NewPos(0, 0)
-		tabBarSize = fyne.NewSize(buttonWidth, size.Height)
-		linePos = fyne.NewPos(buttonWidth, 0)
-		lineSize = fyne.NewSize(theme.Padding(), size.Height)
-		barWidth := buttonWidth + theme.Padding()
-		childPos = fyne.NewPos(barWidth, 0)
-		childSize = fyne.NewSize(size.Width-barWidth, size.Height)
-	case TabLocationBottom:
-		buttonHeight := tabBarMinSize.Height
-		tabBarPos = fyne.NewPos(0, size.Height-buttonHeight)
-		tabBarSize = fyne.NewSize(size.Width, buttonHeight)
-		barHeight := buttonHeight + theme.Padding()
-		linePos = fyne.NewPos(0, size.Height-barHeight)
-		lineSize = fyne.NewSize(size.Width, theme.Padding())
-		childPos = fyne.NewPos(0, 0)
-		childSize = fyne.NewSize(size.Width, size.Height-barHeight)
-	case TabLocationTrailing:
-		buttonWidth := tabBarMinSize.Width
-		tabBarPos = fyne.NewPos(size.Width-buttonWidth, 0)
-		tabBarSize = fyne.NewSize(buttonWidth, size.Height)
-		barWidth := buttonWidth + theme.Padding()
-		linePos = fyne.NewPos(size.Width-barWidth, 0)
-		lineSize = fyne.NewSize(theme.Padding(), size.Height)
-		childPos = fyne.NewPos(0, 0)
-		childSize = fyne.NewSize(size.Width-barWidth, size.Height)
+	// Try render as many tabs as will fit, others will appear in the overflow
```

Comment:
With `MaxAppTabs` this comment is slightly misleading as it will be capped so it's not really as many as will fit.

Reply:
Done